curated_transformers/repository/hf_hub.py

```diff
@@ -31,26 +112,207 @@ def __init__(self, name: str, *, revision: str = "main"):
-                    repo_id=self.name, filename=path, revision=self.revision
-                )
-            )
-        except (
-            EntryNotFoundError,
-            RepositoryNotFoundError,
-            RevisionNotFoundError,
-        ) as e:
-            raise FileNotFoundError(f"File not found: {self.pretty_path(path)}") from e
-        except Exception as e:
-            raise OSError(f"File could not be opened: {self.pretty_path(path)}") from e
+        return HfHubFile(repo=self, path=path)
 
     def pretty_path(self, path: Optional[str] = None) -> str:
         if not path:
             return f"{self.name} (revision: {self.revision})"
         return f"{self.name}/{path} (revision: {self.revision})"
 
+    def transaction(self) -> TransactionContext:
+        return HfHubTransactionContext(self)
+
+
+class HfHubTransactionContext(TransactionContext):
+    """
+    Hugging Face Hub transaction context manager. Files opened
+    for writing are backed by temporary files on the host and
+    uploaded to the Hub when the transaction completes successfully.
+    """
+
+    def __init__(self, repo: HfHubRepository):
+        """
+        :param repo:
+            The parent repository.
+        """
+        super().__init__()
+        self._repo = repo
+        self._file_mappings: Dict[
+            str, IO
+        ] = {}  # Maps remote file paths to local temporary files
+
+    def open(self, path: str, mode: str, encoding: Optional[str] = None) -> IO:
+        if path in self._file_mappings:
+            raise OSError(f"Path '{path}' has already been opened")
+        elif "w" not in mode:
+            raise OSError(
+                f"File open mode '{mode}' is not supported inside a transaction - "
+                "files can only be opened for writing"
+            )
+
+        temp_file = NamedTemporaryFile(mode, encoding=encoding, delete=False)
+        self._file_mappings[path] = temp_file
+        return temp_file
+
+    @property
+    def repo(self) -> Repository:
+        return self._repo
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        try:
+            if exc_type is None:
+                self._upload_temp_files()
+        finally:
+            self._release_temp_files()
+
+    def _upload_temp_files(self):
+        if len(self._file_mappings) == 0:
+            return
+
+        # Flush buffers.
+        for temp_file in self._file_mappings.values():
+            if not temp_file.closed:
+                temp_file.flush()
+
+        # Collate all files as a single commit.
+        commit_ops = [
+            CommitOperationAdd(path_in_repo=remote_path, path_or_fileobj=temp_file.name)
+            for remote_path, temp_file in self._file_mappings.items()
+        ]
+        filenames = "\n".join(self._file_mappings.keys())
+        commit_message = f"Uploading files:\n{filenames}"
+
+        api = HfApi()
+        commit_info = api.create_commit(
```

Comment:
```suggestion
        HfApi().api.create_commit(
```
`commit_info` isn't used, right?

Reply:
True, but I'll `_` it to indicate that we're discarding the result (we might need it later).